agent/src/main/java/com/newrelic/agent/android/NewRelic.java

```diff
@@ -926,13 +927,34 @@ public static boolean removeAllAttributes() {
      * Sets a user ID attribute.
      *
      * @param userId The user ID as string value
-     * @return true if successful, false if the operation did not complete as anticipated.
+     * @return true if userId attribute as created or updated.
      */
     public static boolean setUserId(String userId) {
         StatsEngine.notice().inc(MetricNames.SUPPORTABILITY_API
                 .replace(MetricNames.TAG_NAME, "setUserId"));
 
-        return AnalyticsControllerImpl.getInstance().setAttribute(AnalyticsAttribute.USER_ID_ATTRIBUTE, userId);
+        final AnalyticsControllerImpl controller = AnalyticsControllerImpl.getInstance();
+        final AnalyticsAttribute userIdAttr = controller.getAttribute(AnalyticsAttribute.USER_ID_ATTRIBUTE);
+
+        if (userIdAttr != null) {
+            if (!Objects.equals(userIdAttr.getStringValue(), userId)) {
+                Runnable harvest = () -> {
+                    Harvest.harvestNow(true, true);// call non-blocking harvest
+                    controller.getAttribute(AnalyticsAttribute.SESSION_ID_ATTRIBUTE)
+                            .setStringValue(agentConfiguration.provideSessionId())  // start a new session
+                            .setPersistent(false);
+                     // remove session duration and user id attributes
+                    controller.removeAttribute(AnalyticsAttribute.SESSION_DURATION_ATTRIBUTE);
+                    if (userId == null || userId.isEmpty()) {
+                        controller.removeAttribute(AnalyticsAttribute.USER_ID_ATTRIBUTE);
+                    }
+                };
+                harvest.run();
```

Comment:
This still blocks. You need to submit the runnable to an executor, or create. a new thread and run on that.